rosette/src/Ob.cc

```diff
@@ -76,16 +76,30 @@ int ESCTAG(pOb x) {
     return GET_ESCTAGGED_TAG(te);
 }
 
-int TAGVAL(pOb x) {
+/**
+ * NB(leaf): Originally, this was a macro defined as follows:
+ *
+ * #define SIGN_EXTEND(a, n) ((((int)a) << (WordSize - (n))) >> (WordSize - (n)))
+ *
```

Comment:
IIRC, shift left shifts in 0's from the right (not relevant here), shift right replicates the top bit. So shifting relevant bits to the left edge of the word then back "sign extends" the field. OTOH, I vaguely remember this being implementation dependent. I'm not sure though.